src/ceph_osd.cc

```diff
@@ -472,6 +473,10 @@ int main(int argc, const char **argv)
     forker.exit(0);
   }
   
+  if (extblkdev::preload(g_ceph_context) < 0) {
+    forker.exit(1);
+  }
+
```

Comment:
Valgrind failure looks related to this area:
/a/yuriw-2022-07-22_03:31:37-rados-wip-yuri4-testing-2022-07-21-1832-distro-default-smithi/6943751/remote/smithi123/log/valgrind/osd.5.log.gz
```
    <frame>
      <ip>0x7E6461</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>Preforker::exit(int)</fn>
      <dir>/usr/src/debug/ceph-17.0.0-13730.g2ea4cc97.el8.x86_64/src/common</dir>
      <file>Preforker.h</file>
      <line>132</line>
    </frame>
    <frame>
      <ip>0x79D354</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>main</fn>
      <dir>/usr/src/debug/ceph-17.0.0-13730.g2ea4cc97.el8.x86_64/src</dir>
      <file>ceph_osd.cc</file>
      <line>477</line>
    </frame>
  </stack>
  <auxwhat>Address 0x129baea0 is 416 bytes inside a block of size 1,372,032 free'd</auxwhat>
  <stack>
    <frame>
      <ip>0x4C39A93</ip>
      <obj>/usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so</obj>
      <fn>free</fn>
      <dir>/builddir/build/BUILD/valgrind-3.19.0/coregrind/m_replacemalloc</dir>
      <file>vg_replace_malloc.c</file>
      <line>872</line>
    </frame>
    <frame>
      <ip>0x81DF23B</ip>
      <obj>/usr/lib64/libc-2.28.so</obj>
      <fn>__run_exit_handlers</fn>
    </frame>

...

    <frame>
      <ip>0x12CE723</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>ceph::PluginRegistry::load(std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; const&amp;, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt; const&amp;)</fn>
      <dir>/usr/src/debug/ceph-17.0.0-13730.g2ea4cc97.el8.x86_64/src/common</dir>
      <file>PluginRegistry.cc</file>
      <line>143</line>
    </frame>
    <frame>
      <ip>0x1760394</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>ceph::extblkdev::preload(ceph::common::CephContext*)</fn>
      <dir>/usr/src/debug/ceph-17.0.0-13730.g2ea4cc97.el8.x86_64/src/extblkdev</dir>
      <file>ExtBlkDevPlugin.cc</file>
      <line>176</line>
    </frame>
    <frame>
      <ip>0x79A856</ip>
      <obj>/usr/bin/ceph-osd</obj>
      <fn>main</fn>
      <dir>/usr/src/debug/ceph-17.0.0-13730.g2ea4cc97.el8.x86_64/src</dir>
      <file>ceph_osd.cc</file>
      <line>476</line>
    </frame>
  </stack>
</error>
```

Reply:
The fail indicates that for some reason unclear to me, the new vdo plugin was not found at its location.
The cmake file `src/extblkdev/CMakeLists.txt` defines that location:
```
set(extblkdev_plugin_dir ${CEPH_INSTALL_PKGLIBDIR}/extblkdev)
```
and the cmake file `/home/mohmacht/ceph/src/extblkdev/vdo/CMakeLists.txt` installs it there:
```
add_library(ceph_ebd_vdo SHARED ${vdo_srcs})
target_link_libraries(ceph_ebd_vdo common ${EXTRALIBS})
install(TARGETS ceph_ebd_vdo DESTINATION ${extblkdev_plugin_dir})
```
Also, the `ceph.spec.in` contains a reference to it to ensure it gets added to the RPMs:
```
%dir %{_libdir}/ceph/extblkdev
%{_libdir}/ceph/extblkdev/libceph_*.so*
```

Could you please point me to instructions how to reproduce the runs above, so I can check why the plugin is not found at the default location. Thank you!

Reply:
@mohmacht these runs were done with our teuthology framework (https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/tests-integration-testing-teuthology-intro/). I'm not sure if you have access to it. If not, you can push any new changes to this branch, and a QA engineer (@yuriw) will re-run the tests.

If you do have teuthology access, you can re-test by rerunning any of the tests that failed here against your new changes: http://pulpito.front.sepia.ceph.com/yuriw-2022-07-22_03:31:37-rados-wip-yuri4-testing-2022-07-21-1832-distro-default-smithi/

Reply:
Sorry, I can not create any fix unless I understand the deployment methodology. I browsed through the teuthology documentation, but could not find anything about how the images are created and deployed to the test cluster. The plugin is created with the patch, installed and RPM packaged when running a regular build, but is somehow lost in the deployment methodology used in your cluster. Without recreating the entire path, I can not debug or fix anything...

Reply:
My first guess about what went wrong in the deployment is that a default ceph.spec has been used and not the one generated by this patch. This would explain all the problems.

Reply:
My next guess is that the RPM/ceph.spec path is not even used. I have now added the plugin to the Debian install list that I had missed before.

Reply:
Thanks @mohmacht, I'll have this go through another round of testing. I understand your frustration about teuthology; on my end, I can provide error messages for you to look at, but I'm not sure what can be done about reproducing it on your end.

@aclamk what is usually done in this type of situation?

Reply:
I think best approach would be to
1) get virtual machine with Debian
2) compile ceph
3) make install
It should be the same what is done by teuthology.

Reply:
@aclamk Adam, I just finished your suggestion. I installed a fresh Debian 11, cloned my ceph branch, installed dependencies, ran a vanilla cmake without any options, and ran a make install. Result:
```
mohmacht@debian:~/build$ ls -l /usr/local/lib/ceph/extblkdev/
total 11916
-rw-r--r-- 1 root root 12198464 Aug  3 10:55 libceph_ebd_vdo.so
mohmacht@debian:~/build$ 
```
There must be something else in play here...

Reply:
@ljflores Laura, did my last commit 9bff711 make any difference in integration testing?

Reply:
@mohmacht @aclamk newest test results are here: https://pulpito.ceph.com/yuriw-2022-08-03_14:58:04-rados-wip-yuri4-testing-2022-08-02-1519-distro-default-smithi/

We seem to have gotten rid of the "load failed dlopen" failure in ubuntu, but now we are hitting a segfault in many tests regardless of distro:

```
2022-08-03T15:28:32.314 INFO:tasks.ceph.osd.1:Restarting daemon
2022-08-03T15:28:32.315 DEBUG:teuthology.orchestra.run.smithi086:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage daemon-helper kill ceph-osd -f --cluster ceph -i 1
2022-08-03T15:28:32.317 INFO:tasks.ceph.osd.1:Started
2022-08-03T15:28:32.318 DEBUG:teuthology.orchestra.run.smithi086:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph osd dump --format=json
2022-08-03T15:28:32.404 INFO:tasks.ceph.osd.1.smithi086.stderr:*** Caught signal (Segmentation fault) **
2022-08-03T15:28:32.404 INFO:tasks.ceph.osd.1.smithi086.stderr: in thread 7fe3236da500 thread_name:ceph-osd
2022-08-03T15:28:32.494 INFO:teuthology.orchestra.run.smithi086.stdout:
2022-08-03T15:28:32.494 INFO:teuthology.orchestra.run.smithi086.stdout:{"epoch":4,"fsid":"11d60c2f-0d4c-4e31-a42e-b748267f2ccf","created":"2022-08-03T15:28:31.497834+0000","modified":"2022-08-03T15:28:32.296619+0000","last_up_change":"0.000000","last_in_change":"2022-08-03T15:28:32.296619+0000","flags":"sortbitwise,recovery_deletes,purged_snapdirs,pglog_hardlimit","flags_num":5799936,"flags_set":["pglog_hardlimit","purged_snapdirs","recovery_deletes","sortbitwise"],"crush_version":2,"full_ratio":0.94999998807907104,"backfillfull_ratio":0.89999997615814209,"nearfull_ratio":0.85000002384185791,"cluster_snapshot":"","pool_max":0,"max_osd":2,"require_min_compat_client":"luminous","min_compat_client":"jewel","require_osd_release":"quincy","pools":[],"osds":[{"osd":0,"uuid":"8401ce52-f069-4678-90a4-fb0f1474b436","up":0,"in":1,"weight":1,"primary_affinity":1,"last_clean_begin":0,"last_clean_end":0,"up_from":0,"up_thru":0,"down_at":0,"lost_at":0,"public_addrs":{"addrvec":[]},"cluster_addrs":{"addrvec":[]},"heartbeat_back_addrs":{"addrvec":[]},"heartbeat_front_addrs":{"addrvec":[]},"public_addr":"(unrecognized address family 0)/0","cluster_addr":"(unrecognized address family 0)/0","heartbeat_back_addr":"(unrecognized address family 0)/0","heartbeat_front_addr":"(unrecognized address family 0)/0","state":["exists","new"]},{"osd":1,"uuid":"4487c011-05ae-484b-845b-1bc0c7d5fe77","up":0,"in":1,"weight":1,"primary_affinity":1,"last_clean_begin":0,"last_clean_end":0,"up_from":0,"up_thru":0,"down_at":0,"lost_at":0,"public_addrs":{"addrvec":[]},"cluster_addrs":{"addrvec":[]},"heartbeat_back_addrs":{"addrvec":[]},"heartbeat_front_addrs":{"addrvec":[]},"public_addr":"(unrecognized address family 0)/0","cluster_addr":"(unrecognized address family 0)/0","heartbeat_back_addr":"(unrecognized address family 0)/0","heartbeat_front_addr":"(unrecognized address family 0)/0","state":["exists","new"]}],"osd_xinfo":[{"osd":0,"down_stamp":"0.000000","laggy_probability":0,"laggy_interval":0,"features":0,"old_weight":0,"last_purged_snaps_scrub":"0.000000","dead_epoch":0},{"osd":1,"down_stamp":"0.000000","laggy_probability":0,"laggy_interval":0,"features":0,"old_weight":0,"last_purged_snaps_scrub":"0.000000","dead_epoch":0}],"pg_upmap":[],"pg_upmap_items":[],"pg_temp":[],"primary_temp":[],"blocklist":{},"range_blocklist":{},"erasure_code_profiles":{"default":{"crush-failure-domain":"osd","k":"2","m":"1","plugin":"jerasure","technique":"reed_sol_van"}},"removed_snaps_queue":[],"new_removed_snaps":[],"new_purged_snaps":[],"crush_node_flags":{},"device_class_flags":{},"stretch_mode":{"stretch_mode_enabled":false,"stretch_bucket_count":0,"degraded_stretch_mode":0,"recovering_stretch_mode":0,"stretch_mode_bucket":0}}
2022-08-03T15:28:32.506 INFO:tasks.ceph.ceph_manager.ceph:[]
2022-08-03T15:28:32.507 INFO:tasks.ceph:Waiting for OSDs to come up
2022-08-03T15:28:32.560 INFO:tasks.ceph.osd.0.smithi086.stderr:daemon-helper: command failed with exit status 1
2022-08-03T15:28:32.568 INFO:tasks.ceph.osd.1.smithi086.stderr:daemon-helper: command crashed with signal 11
2022-08-03T15:28:32.570 ERROR:teuthology.contextutil:Saw exception from nested tasks
```

Inspecting the coredump with gdb reveals:
```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `ceph-osd -f --cluster ceph -i 1'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fe325012fd3 in tcmalloc::ThreadCache::ReleaseToCentralCache(tcmalloc::ThreadCache::FreeList*, unsigned int, int) () from /lib/x86_64-linux-gnu/libtcmalloc.so.4
[Current thread is 1 (Thread 0x7fe3236da500 (LWP 23607))]
(gdb) bt
#0  0x00007fe325012fd3 in tcmalloc::ThreadCache::ReleaseToCentralCache(tcmalloc::ThreadCache::FreeList*, unsigned int, int) () from /lib/x86_64-linux-gnu/libtcmalloc.so.4
#1  0x00007fe325013350 in tcmalloc::ThreadCache::ListTooLong(tcmalloc::ThreadCache::FreeList*, unsigned int) () from /lib/x86_64-linux-gnu/libtcmalloc.so.4
#2  0x0000556ca0df9333 in ceph::ClibBackTrace::demangle[abi:cxx11](char const*) (name=0x7ffe25dc1050 "\360Tb\244lU") at ./src/common/BackTrace.cc:61
#3  0x0000556ca0df95b5 in ceph::ClibBackTrace::print (this=this@entry=0x7ffe25dc11c0, out=...) at /usr/include/c++/11/ostream:606
#4  0x0000556ca0ded9d5 in handle_oneshot_fatal_signal (signum=11) at ./src/global/signal_handler.cc:327
#5  <signal handler called>
#6  0x00007fe325012fd3 in tcmalloc::ThreadCache::ReleaseToCentralCache(tcmalloc::ThreadCache::FreeList*, unsigned int, int) () from /lib/x86_64-linux-gnu/libtcmalloc.so.4
#7  0x00007fe325013350 in tcmalloc::ThreadCache::ListTooLong(tcmalloc::ThreadCache::FreeList*, unsigned int) () from /lib/x86_64-linux-gnu/libtcmalloc.so.4
#8  0x0000556ca0e9cb6a in __gnu_cxx::new_allocator<char>::deallocate (this=<optimized out>, __t=<optimized out>, __p=<optimized out>) at /usr/include/c++/11/ext/new_allocator.h:126
#9  std::allocator<char>::deallocate (__n=<optimized out>, __p=<optimized out>, this=<optimized out>) at /usr/include/c++/11/bits/allocator.h:187
#10 std::allocator_traits<std::allocator<char> >::deallocate (__n=<optimized out>, __p=<optimized out>, __a=...) at /usr/include/c++/11/bits/alloc_traits.h:492
#11 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_destroy (__size=<optimized out>, this=<optimized out>) at /usr/include/c++/11/bits/basic_string.h:237
#12 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_dispose (this=<optimized out>) at /usr/include/c++/11/bits/basic_string.h:232
#13 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string (this=<optimized out>, __in_chrg=<optimized out>) at /usr/include/c++/11/bits/basic_string.h:658
#14 Option::~Option (this=<optimized out>, __in_chrg=<optimized out>) at ./src/common/options.h:14
#15 std::destroy_at<Option> (__location=<optimized out>) at /usr/include/c++/11/bits/stl_construct.h:88
#16 std::_Destroy<Option> (__pointer=<optimized out>) at /usr/include/c++/11/bits/stl_construct.h:138
#17 std::_Destroy_aux<false>::__destroy<Option*> (__last=<optimized out>, __first=0x556ca47e1b40) at /usr/include/c++/11/bits/stl_construct.h:152
#18 std::_Destroy<Option*> (__last=<optimized out>, __first=<optimized out>) at /usr/include/c++/11/bits/stl_construct.h:185
#19 std::_Destroy<Option*, Option> (__last=<optimized out>, __first=<optimized out>) at /usr/include/c++/11/bits/alloc_traits.h:746
#20 std::vector<Option, std::allocator<Option> >::~vector (this=0x556ca20c7190 <ceph_options>, __in_chrg=<optimized out>) at /usr/include/c++/11/bits/stl_vector.h:680
#21 0x00007fe3247308a7 in __run_exit_handlers (status=1, listp=0x7fe3248d6718 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:108
#22 0x00007fe324730a60 in __GI_exit (status=<optimized out>) at exit.c:139
#23 0x0000556ca06bfd28 in Preforker::exit (this=<optimized out>, r=1) at ./src/common/Preforker.h:132
#24 0x0000556ca068f826 in main (argc=<optimized out>, argv=<optimized out>) at ./src/ceph_osd.cc:477
```

Reply:
It looks like these tests revealed a race condition where a string was modified concurrently by multiple threads and led to a double free. The message points to strings in the options storage, but the only string access to this is
```
      string plugins = conf.get_val<std::string>("osd_extblkdev_plugins");
```
and this call should be thread safe.
I added locks to plugin registry queries introduced by this patch (commit 1d39740dc13), it is possible that concurrent querying of plugins and deletion of the entire plugin registry caused the race.
@ljflores Laura, could you please rerun with this patch? Thank you!

Reply:
@mohmacht rerunning... will update with new results when available.

Reply:
@mohmacht actually, first see the failure in make check, as it is related to your changes:

```
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/standalone/ceph-helpers.sh:680: run_osd:  grep -q '"noup"'
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/standalone/ceph-helpers.sh:680: run_osd:  jq '.flags_set[]'
2022-08-04T17:51:21.139+0000 7fa835b26380 -1 WARNING: all dangerous and experimental features are enabled.
2022-08-04T17:51:21.199+0000 7fa835b26380 -1 WARNING: all dangerous and experimental features are enabled.
2022-08-04T17:51:21.219+0000 7fa835b26380 -1 WARNING: all dangerous and experimental features are enabled.
terminate called after throwing an instance of 'std::system_error'
  what():  Resource deadlock avoided
*** Caught signal (Aborted) **
 in thread 7fa835b26380 thread_name:ceph-osd
 ceph version Development (no_version) quincy (dev)
 1: ceph-osd() [0x33f53a5]
 2: /lib/x86_64-linux-gnu/libpthread.so.0(+0x153c0) [0x7fa8373e63c0]
 3: gsignal()
 4: abort()
 5: /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xa1a31) [0x7fa836d38a31]
 6: /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xad5dc) [0x7fa836d445dc]
 7: /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xad647) [0x7fa836d44647]
 8: /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0xad8e9) [0x7fa836d448e9]
 9: (ceph::mutex_debug_detail::mutex_debug_impl<false>::lock_impl()+0xaa) [0x26a1eba]
 10: (ceph::mutex_debug_detail::mutex_debug_impl<false>::lock(bool)+0x74) [0x26a1d04]
 11: (std::lock_guard<ceph::mutex_debug_detail::mutex_debug_impl<false> >::lock_guard(ceph::mutex_debug_detail::mutex_debug_impl<false>&)+0x27) [0x274b527]
 12: (ceph::extblkdev::get_required_caps(ceph::common::CephContext*, _cap_struct*&)+0xaf) [0x4323a8f]
 13: (ceph::extblkdev::limit_caps(ceph::common::CephContext*)+0x84) [0x43243a4]
 14: (ceph::extblkdev::preload(ceph::common::CephContext*)+0x2e2) [0x4324772]
 15: main()
 16: __libc_start_main()
 17: _start()
2022-08-04T17:51:21.443+0000 7fa835b26380 -1 *** Caught signal (Aborted) **
 in thread 7fa835b26380 thread_name:ceph-osd
```

You can check the full log either by clicking on the make check details in the checks that were run, or by clicking here: https://jenkins.ceph.com/job/ceph-pull-requests/100847/consoleFull

Once this is resolved, I can rerun integration tests.

Reply:
Sorry about that, I should have caught that before committing... Fixing now...

Reply:
The 0543d29 commit just passed the run-make-check.sh suite.

Reply:
@mohmacht rerunning your changes...

Reply:
@mohmacht we are still hitting segfaults. The problem occurs with the `forker.exit(1)` line in this file.
```
Core was generated by `ceph-osd -f --cluster ceph -i 0'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f0fd46a3fd3 in tcmalloc::ThreadCache::ReleaseToCentralCache(tcmalloc::ThreadCache::FreeList*, unsigned int, int) () from /lib/x86_64-linux-gnu/libtcmalloc.so.4
[Current thread is 1 (Thread 0x7f0fd2d6b500 (LWP 30281))]
(gdb) bt
#0  0x00007f0fd46a3fd3 in tcmalloc::ThreadCache::ReleaseToCentralCache(tcmalloc::ThreadCache::FreeList*, unsigned int, int) () from /lib/x86_64-linux-gnu/libtcmalloc.so.4
#1  0x00007f0fd46a4350 in tcmalloc::ThreadCache::ListTooLong(tcmalloc::ThreadCache::FreeList*, unsigned int) () from /lib/x86_64-linux-gnu/libtcmalloc.so.4
#2  0x000055e7afef06a3 in ceph::ClibBackTrace::demangle[abi:cxx11](char const*) ()
#3  0x000055e7afef0925 in ceph::ClibBackTrace::print(std::ostream&) const ()
#4  0x000055e7afee4d45 in ?? ()
#5  <signal handler called>
#6  0x00007f0fd46a3fd3 in tcmalloc::ThreadCache::ReleaseToCentralCache(tcmalloc::ThreadCache::FreeList*, unsigned int, int) () from /lib/x86_64-linux-gnu/libtcmalloc.so.4
#7  0x00007f0fd46a4350 in tcmalloc::ThreadCache::ListTooLong(tcmalloc::ThreadCache::FreeList*, unsigned int) () from /lib/x86_64-linux-gnu/libtcmalloc.so.4
#8  0x000055e7aff93eca in std::vector<Option, std::allocator<Option> >::~vector() ()
#9  0x00007f0fd3dc18a7 in __run_exit_handlers (status=1, listp=0x7f0fd3f67718 <__exit_funcs>, run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:108
#10 0x00007f0fd3dc1a60 in __GI_exit (status=<optimized out>) at exit.c:139
#11 0x000055e7af7b6f18 in Preforker::exit(int) ()
#12 0x000055e7af7868b6 in main ()
```

As you can see from the log, these segfaults are hit when the OSD daemons are restarted:
```
2022-08-05T15:15:06.683 INFO:tasks.ceph.osd.4:Restarting daemon
2022-08-05T15:15:06.684 DEBUG:teuthology.orchestra.run.smithi185:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage daemon-helper kill ceph-osd -f --cluster ceph -i 4
2022-08-05T15:15:06.685 INFO:tasks.ceph.osd.0.smithi185.stderr:*** Caught signal (Segmentation fault) **
2022-08-05T15:15:06.703 INFO:tasks.ceph.osd.0.smithi185.stderr: in thread 7f0fd2d6b500 thread_name:ceph-osd
2022-08-05T15:15:06.705 INFO:tasks.ceph.osd.1.smithi185.stderr:*** Caught signal (Segmentation fault) **
2022-08-05T15:15:06.705 INFO:tasks.ceph.osd.1.smithi185.stderr: in thread 7f23fe8f6500 thread_name:ceph-osd
2022-08-05T15:15:06.707 INFO:tasks.ceph.osd.4:Started
2022-08-05T15:15:06.707 INFO:tasks.ceph.osd.5:Restarting daemon
2022-08-05T15:15:06.708 DEBUG:teuthology.orchestra.run.smithi185:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage daemon-helper kill ceph-osd -f --cluster ceph -i 5
2022-08-05T15:15:06.712 INFO:tasks.ceph.osd.5:Started
2022-08-05T15:15:06.713 INFO:tasks.ceph.osd.6:Restarting daemon
2022-08-05T15:15:06.714 DEBUG:teuthology.orchestra.run.smithi185:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage daemon-helper kill ceph-osd -f --cluster ceph -i 6
2022-08-05T15:15:06.724 INFO:tasks.ceph.osd.6:Started
2022-08-05T15:15:06.725 INFO:tasks.ceph.osd.7:Restarting daemon
2022-08-05T15:15:06.725 DEBUG:teuthology.orchestra.run.smithi185:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage daemon-helper kill ceph-osd -f --cluster ceph -i 7
2022-08-05T15:15:06.747 INFO:tasks.ceph.mgr.x.smithi185.stderr:2022-08-05T15:15:06.719+0000 7f5855b8d140 -1 mgr[py] Module prometheus has missing NOTIFY_TYPES member
2022-08-05T15:15:06.758 INFO:tasks.ceph.osd.2.smithi185.stderr:*** Caught signal (Segmentation fault) **
2022-08-05T15:15:06.783 INFO:tasks.ceph.osd.2.smithi185.stderr: in thread 7f9fbc6eb500 thread_name:ceph-osd
2022-08-05T15:15:06.810 INFO:tasks.ceph.osd.3.smithi185.stderr:*** Caught signal (Segmentation fault) **
2022-08-05T15:15:06.811 INFO:tasks.ceph.osd.3.smithi185.stderr: in thread 7fa8e4f20500 thread_name:ceph-osd
2022-08-05T15:15:06.814 INFO:tasks.ceph.osd.5.smithi185.stderr:*** Caught signal (Segmentation fault) **
2022-08-05T15:15:06.814 INFO:tasks.ceph.osd.5.smithi185.stderr: in thread 7f87b810b500 thread_name:ceph-osd
2022-08-05T15:15:06.815 INFO:tasks.ceph.osd.4.smithi185.stderr:*** Caught signal (Segmentation fault) **
2022-08-05T15:15:06.817 INFO:tasks.ceph.osd.4.smithi185.stderr: in thread 7f0e2a890500 thread_name:ceph-osd
2022-08-05T15:15:06.819 INFO:tasks.ceph.osd.6.smithi185.stderr:*** Caught signal (Segmentation fault) **
2022-08-05T15:15:06.820 INFO:tasks.ceph.osd.6.smithi185.stderr: in thread 7fe1f5917500 thread_name:ceph-osd
```

Are you able to produce this scenario locally with your branch? I'd try restarting the OSDs to try and recreate the issue. I'll say also that these segfaults are occurring on a variety of different tests, so it's not just once testing scenario that you'd need to replicate. This seems to be reproducible in a variety of scenarios by restarting the OSD.

@aclamk any suggestions?

Reply:
@ljflores Laura, I tried to reproduce the fail by shutting down OSDs started with start by either shooting them down them with `kill` or by `ceph osd stop`, but all these shutdowns were clean. Do you know how the OSDs were restarted? Or how I can get access to your pulpit machine to examine the fails more closely?

Reply:
Please disregard my request for access to pulpito, somehow I can access it today...

Reply:
@ljflores Hi Laura, I think I tracked the problem down to a library reference in the plugin that caused a library to be loaded rather than using the one already loaded by ceph-osd itself. The fix is in fe5b747f61. Could you please restart the integration test? Thank you!

Reply:
@mohmacht I'd like another review from @aclamk first before putting this through integration tests. @aclamk thoughts?

Reply:
@ljflores Laura, could you please just run one small batch of integration tests just to validate (or refute) the commit I pushed last week? Thank you...

Reply:
Building here... https://shaman.ceph.com/builds/ceph/mohmacht-ebdplugin/b4822f4933c990084b4b1f766ac511c08aa4a45a/